zcash_extras/src/wallet.rs

```diff
@@ -2,23 +2,62 @@
 use std::fmt::{Debug, Display};
 
 use zcash_primitives::{
-    consensus::{self, BranchId},
+    consensus::{self, BranchId, NetworkUpgrade},
     memo::MemoBytes,
     sapling::prover::TxProver,
     transaction::{
         builder::Builder,
         components::{amount::DEFAULT_FEE, Amount},
+        Transaction,
     },
     zip32::{ExtendedFullViewingKey, ExtendedSpendingKey},
 };
 
 use crate::WalletWrite;
 use zcash_client_backend::{
     address::RecipientAddress,
-    data_api::{error::Error, SentTransaction},
+    data_api::{error::Error, ReceivedTransaction, SentTransaction},
+    decrypt_transaction,
     wallet::{AccountId, OvkPolicy},
 };
 
+/// Scans a [`Transaction`] for any information that can be decrypted by the accounts in
+/// the wallet, and saves it to the wallet.
+pub async fn decrypt_and_store_transaction<N, E, P, D>(
+    params: &P,
+    data: &mut D,
+    tx: &Transaction,
+) -> Result<(), E>
+where
+    E: From<Error<N>>,
+    P: consensus::Parameters,
+    D: WalletWrite<Error = E>,
+{
+    // Fetch the ExtendedFullViewingKeys we are tracking
+    let extfvks = data.get_extended_full_viewing_keys().await?;
+
+    let max_height = data.block_height_extrema().await?.map(|(_, max)| max + 1);
+    let height = data
+        .get_tx_height(tx.txid())
+        .await?
+        .or(max_height)
```

Comment:
For unconfirmed transactions which this PR is for, will using the max height cause problems in `decrypt_transaction`? `get_tx_height` will return nothing as it's not in db so max height will be used.

Reply:
except from zcash protocol spec.
> • [Heartwood onward] All Sapling and Orchard outputs in coinbase transactions MUST decrypt to a note
plaintext, i.e. the procedure in § 4.20.3 ‘Decryption using a Full Viewing Key (Sapling Sapling Sapling Sapling and Orchard Orchard Orchard Orchard)’ on p. 68 does not return ⊥, using a sequence of 32 zero bytes as the outgoing viewing key. (This implies that before Canopy activation Sapling outputs of a coinbase transaction MUST have note plaintext lead byte equal to 0x01.)

So using max_height would only cause issues during the `Canopy` network upgrade window
https://github.com/KomodoPlatform/librustzcash/blob/e92443a7bbd1c5e92e00e6deb45b5a33af14cea4/zcash_primitives/src/sapling/note_encryption.rs#L322-L338

albeit we still need to test...

Reply:
As discussed in DM, it makes sense to use the max height since it's the next anticipated block which we wish the transaction to be in. We should remove this though https://github.com/KomodoPlatform/librustzcash/blob/6aa3917e95aaac05f0a81be325dcbab39337b0cc/zcash_extras/src/wallet.rs#L44-L45 and use `unwrap_or_else` instead of map here https://github.com/KomodoPlatform/librustzcash/blob/6aa3917e95aaac05f0a81be325dcbab39337b0cc/zcash_extras/src/wallet.rs#L39

Reply:
> We should remove this though 

Why? I mean height will be optional which isn't what we want.

Reply:
ok that makes sense

Reply:
> So using max_height would only cause issues during the Canopy network upgrade window 

AFAIK we do not support zcash privacy transactions (yet), right?
Only supporting PIRATE which is still on Sapling.

Reply:
Yep.. I mean shielded transaction.